### PR TITLE
[Semantics] Use a context object for giving better errors

### PR DESCRIPTION
**Pull request by c-harding**
This is in preparation for the validator to use the same contextual error handler

### CODE REVIEW

File: src/main/scala/temple/DSL/semantics/Analyzer.scala [hunk 1]
```diff
@@ -78,65 +78,64 @@ object Analyzer {
...
+              argMap.getOptionArg("min", IntArgType)(innerContext),
+              argMap.getOptionArg("max", IntArgType)(innerContext),
+              argMap.getArg("precision", IntArgType)(innerContext).toByte,
             )
           case "string" =>
             val argMap = parseParameters(
               "maxLength" -> Some(Arg.NoArg),
               "minLength" -> Some(Arg.NoArg),
-            )(args)
+            )(args)(innerContext)
```
Comment (jaylees14) [inline src/main/scala/temple/DSL/semantics/Analyzer.scala:91]: Much curry

**Review by jaylees14 — COMMENTED** [review]
LGTM, would be good to see some testing? Or potentially manual examples of how to test it?

File: src/main/scala/temple/DSL/semantics/Analyzer.scala (hunk 1 above)
Comment (Lewiky) [reply src/main/scala/temple/DSL/semantics/Analyzer.scala:91]: Like Birmingham in here

File: src/main/scala/temple/DSL/semantics/Context.scala [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+  def error(msg: String): SemanticParsingException = new SemanticParsingException(s"$msg in $this")
+  def fail(msg: String): Nothing                   = throw error(msg)
+
+  def apply[T](f: T => Context => Unit)(name: String, t: T): Unit = f(t)(this :+ name)
+}
+
+object Context {
+  val empty: Context             = Context(List())
+  def from(xs: String*): Context = xs.foldLeft(empty)(_ :+ _)
```
Comment (Lewiky) [inline src/main/scala/temple/DSL/semantics/Context.scala]: This is witchcraft

**Review by Lewiky — APPROVED** [review]
Looks good - some pure black magic going on here but I'm all for it

File: src/main/scala/temple/DSL/semantics/MetadataParser.scala [hunk 3]
```diff
@@ -16,52 +17,80 @@ import scala.collection.mutable
...
+        MetadataParser.assertNoParameters(args)
         constructor
       })
 
   /** Perform parsing by looking up the relevant function and passing it the argument list */
-  final def apply(metaKey: String, args: Args)(implicit context: BlockContext): T =
-    matchers.get(metaKey).map(_(args)) getOrElse {
-      fail(s"No valid metadata $metaKey in $context")
+  final def apply(metaKey: String, args: Args)(implicit context: SemanticContext): T =
+    matchers.get(metaKey).map(_(args, context)) getOrElse {
```
Comment (jaylees14) [inline src/main/scala/temple/DSL/semantics/MetadataParser.scala]: Maybe make this more explicit

**Review by jaylees14 — APPROVED** [review]

**Pull request merged**